src/utils/AutoControlledComponent-slim.js

```diff
@@ -0,0 +1,45 @@
+/* eslint-disable */
+import _ from 'lodash'
+import { Component } from 'react'
+
+const toDefault = (prop) => `default${prop[0].toUpperCase() + prop.slice(1)}`
+
+export default class AutoControlledComponent extends Component {
+  componentWillMount() {
+    if (super.componentWillMount) super.componentWillMount()
+    const { autoControlledProps } = this.constructor
+
+    this.state = _.transform(autoControlledProps, (res, prop) => {
+      const defaultProp = toDefault(prop)
+
+      // apply default props if they exist
+      res[prop] = this.props[defaultProp in this.props ? defaultProp : prop]
+    }, {})
+  }
+
+  componentWillReceiveProps(nextProps) {
+    super.componentWillReceiveProps(nextProps)
+
+    // props always win, update state with all auto controlled prop
+    const newState = _.pick(nextProps, this.constructor.autoControlledProps)
+    if (!_.isEmpty(newState)) this.setState(newState)
+  }
+
+  /**
+   * Safely attempt to set state for props that might be controlled by the user.
+   * Second argument is a state object that is always passed to setState.
+   * @param {object} maybeState State that corresponds to controlled props.
+   * @param {object} [state] Actual state, useful when you also need to setState.
```

Comment:
We solved this offline already, for the record, this http://www.react.run/4kJFdKoxb/27 shows how the single vs double render happens when using vanilla DOM events vs React events.

Since we have to have DOM events for the "click outside" and other features, we're in the pickle.